lib/AST/Type.cpp

```diff
@@ -1230,6 +1230,22 @@ bool ExistentialLayout::containsNonMarkerProtocols() const {
   return false;
 }
 
+bool ExistentialLayout::isABICompatibleWith(
```

Comment:
I'm not sure I follow. There are other situations where two existential types are ABI compatible but not the same type in the compiler, for example `P & AnyObject` vs `P & C` for an arbitrary class `C`. Does the compiler also enter an infinite loop in that case?